nighthawk/source/client/stream_decoder.cc

```diff
@@ -0,0 +1,63 @@
+#include "nighthawk/source/client/stream_decoder.h"
+
+#include <memory>
+
+#include "common/http/http1/codec_impl.h"
+#include "common/http/utility.h"
+
+namespace Nighthawk {
+namespace Client {
+
+void StreamDecoder::decodeHeaders(Envoy::Http::HeaderMapPtr&& headers, bool end_stream) {
+  ASSERT(!complete_);
+  complete_ = end_stream;
+  response_headers_ = std::move(headers);
+  if (complete_) {
+    onComplete(true);
+  }
+}
+
+void StreamDecoder::decodeData(Envoy::Buffer::Instance&, bool end_stream) {
+  ASSERT(!complete_);
+  complete_ = end_stream;
+  if (complete_) {
+    onComplete(true);
+  }
+}
+
+void StreamDecoder::decodeTrailers(Envoy::Http::HeaderMapPtr&&) {
+  ASSERT(!complete_);
+  complete_ = true;
+  onComplete(true);
+}
+
+void StreamDecoder::onComplete(bool success) {
+  if (success && measure_latencies_) {
+    latency_statistic_.addValue((time_source_.monotonicTime() - request_start_).count());
+  }
+  ASSERT(!success || complete_);
+  decoder_completion_callback_.onComplete(success, *response_headers_);
+  if (success) {
```

Comment:
Should we be computing success/failure stats?

Reply:
`BenchmarkClientHttpImpl` will do that for us:
https://github.com/envoyproxy/envoy-perf/pull/38/files#diff-28811197472bd6ba614ce0f99b28024fR210

Also, we leverage Envoy:

https://github.com/envoyproxy/envoy-perf/pull/38/files#diff-547dca47cd8b31225f4d73460ac4af1eR234

In the future, we could also consider plugging in something that aggregates counts per response status or group (4xx, 5xx). We could make that generic by having a function inspect the response headers, and returning a stat name that should be incremented?

Reply:
SG